src/groupby/groupby.cuh

```diff
@@ -0,0 +1,418 @@
+                                                      out_groupby_columns, out_aggregation_column, sort_result);
+      }
+    case GDF_INT32:  
+      { 
+        return dispatch_aggregation_type<int32_t, op>(ncols, in_groupby_columns, in_aggregation_column, 
+                                                      out_groupby_columns, out_aggregation_column, sort_result);
+      }
+    case GDF_INT64:  
+      { 
+        return dispatch_aggregation_type<int64_t, op>(ncols, in_groupby_columns, in_aggregation_column, 
+                                                      out_groupby_columns, out_aggregation_column, sort_result);
+      }
+    // For floating point groupby column types, cast to an integral type
+    case GDF_FLOAT32:
+      {
+        return dispatch_aggregation_type<int32_t, op>(ncols, in_groupby_columns, in_aggregation_column, 
+                                                      out_groupby_columns, out_aggregation_column, sort_result);
+      }
+    case GDF_FLOAT64:
+      { 
+        return dispatch_aggregation_type<int64_t, op>(ncols, in_groupby_columns, in_aggregation_column, 
+                                                      out_groupby_columns, out_aggregation_column, sort_result);
+      }
+    default:
+     {
+       std::cout << "Unsupported groupby column type:" << groupby_column_type << std::endl;
+       return GDF_UNSUPPORTED_DTYPE;
+     }
+  }
+
+}
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @Synopsis  This function provides the libgdf entry point for a hash-based group-by.
+ * Performs a Group-By operation on an arbitrary number of columns with a single aggregation column.
+ * 
+ * @Param[in] ncols The number of columns to group-by
+ * @Param[in] in_groupby_columns[] The columns to group-by
+ * @Param[in,out] in_aggregation_column The column to perform the aggregation on
+ * @Param[in,out] out_groupby_columns[] A preallocated buffer to store the resultant group-by columns
+ * @Param[in,out] out_aggregation_column A preallocated buffer to store the resultant aggregation column
+ * @tparam[in] aggregation_operation A functor that defines the aggregation operation
+ * 
+ * @Returns gdf_error
+ */
+/* ----------------------------------------------------------------------------*/
+template <template <typename aggregation_type> class aggregation_operation>
+gdf_error gdf_group_by_hash(int ncols,               
+                            gdf_column* in_groupby_columns[],        
+                            gdf_column* in_aggregation_column,       
+                            gdf_column* out_groupby_columns[],
+                            gdf_column* out_aggregation_column,
+                            bool sort_result = false)
+{
+
+  // TODO Currently only supports a single groupby column
+  if(ncols > 1) {
+    assert( false && "Can only support a single groupby column at this time.");
```

Comment:
Is this a non-standard `assert()` macro?

Reply:
Yeah, that's a remnant from my hacky C days. I replaced it with returning a new error code if a groupby is attempted with more than one column.